test/sync/change.test.ts

```diff
@@ -87,4 +114,466 @@ describe("sync/change", () => {
+        });
+
+        it("git show throws on delete", async () => {
+            const fileContents = await prv.previousSpecVersion("", "", "");
+            assert.equal(fileContents, "");
+        });
+    });
+
+    describe("changeResources", () => {
+
+        it("resource file does not exist", async () => {
+            const project: GitProject = InMemoryProject.of() as any;
+            const diff: PushDiff = {
+                change: "apply",
+                path: "fake.path",
+                sha: "fake.sha",
+            };
+
+            try {
+                await changeResource(project, diff);
+                assert.fail("should not", "get here");
+            } catch (e) {
+                assert.equal(e.message, "Resource spec file 'fake.path' does not exist in project");
+            }
+        });
+
+        describe("spoofs", () => {
+            const resource = {
+                apiVersion: "v1",
+                kind: "Secret",
+                type: "Opaque",
+                metadata: {
+                    name: "mysecret",
+                },
+                data: {
+                    username: "dGhlIHJvb3RtaW5pc3RyYXRvcg==",
+                    password: "Y29ycmVjdCBob3JzZSBiYXR0ZXJ5IHN0YXBsZQ==",
+                },
+            };
+
+            let origClientDelete: any;
+            let origClientPatch: any;
+            let origClientRead: any;
+            let origPreviousSpecVersion: any;
+
+            before(() => {
+                origClientDelete = Object.getOwnPropertyDescriptor(api.K8sObjectApi.prototype, "delete");
+                origClientPatch = Object.getOwnPropertyDescriptor(api.K8sObjectApi.prototype, "patch");
+
+                origClientRead = Object.getOwnPropertyDescriptor(api.K8sObjectApi.prototype, "read");
+                Object.defineProperty(api.K8sObjectApi.prototype, "read", {
+                    value: async (spec: k8s.KubernetesObject) => {
+                        return Promise.resolve();
+                    },
+                });
+
+                origPreviousSpecVersion = Object.getOwnPropertyDescriptor(prv, "previousSpecVersion");
+            });
+
+            after(() => {
+                Object.defineProperty(api.K8sObjectApi.prototype, "delete", origClientDelete);
+                Object.defineProperty(api.K8sObjectApi.prototype, "patch", origClientPatch);
+                Object.defineProperty(api.K8sObjectApi.prototype, "read", origClientRead);
+                Object.defineProperty(prv, "previousSpecVersion", origPreviousSpecVersion);
+            });
+
+            /**
+             * expect that the patch method is not called
+             */
+            it("delete changes", async () => {
+                Object.defineProperty(api.K8sObjectApi.prototype, "delete", {
+                    value: async (spec: k8s.KubernetesObject, body: any) => {
+                        return Promise.resolve();
```

Comment:
It can be helpful to set a variable value here to ensure this function gets called.

```typescript
                let deleteCalled = false;
                Object.defineProperty(api.K8sObjectApi.prototype, "delete", {
                    value: async (spec: k8s.KubernetesObject, body: any) => {
                        deleteCalled = true;
                        return Promise.resolve();
                    },
                });
                …
                assert(deleteCalled, "delete was never called");
```

Reply:
makes sense, will use that pattern for a couple of the tests